src/astarte-client/client.ts

```diff
@@ -320,17 +331,48 @@ class AstarteClient {
     deviceId: AstarteDevice['id'];
     interfaceName: AstarteInterface['name'];
   }): Promise<AstarteInterfaceValues> {
-    const { deviceId, interfaceName } = params;
     const response = await this.$get(
       this.apiConfig.deviceData({
-        deviceId,
-        interfaceName,
+        deviceId: params.deviceId,
+        interfaceName: params.interfaceName,
         ...this.config,
       }),
     );
     return response.data;
   }
 
+  async getDeviceDataTree(
+    params: { deviceId: AstarteDevice['id'] } & InterfaceOrInterfaceNameParams,
+  ): Promise<
+    | AstarteDataTreeRoot<AstartePropertyData>
+    | AstarteDataTreeRoot<AstarteDatastreamIndividualData>
+    | AstarteDataTreeRoot<AstarteDatastreamObjectData>
+    | AstarteDataTreeLeaf<AstarteDatastreamObjectData>
```

Comment:
Now I am returning
```
| AstarteDataTree<AstartePropertyData>
| AstarteDataTree<AstarteDatastreamIndividualData>
| AstarteDataTree<AstarteDatastreamObjectData>
```
The user isn't forced to cast or check for leaves, but he can be sure to have the leaves with `getLeaves()`.
However there's a difference between a branch and a leaf, and it could be needlessly tricky to hide that difference to the user: for instance, `branch.data` returns an array of leaves data, `leaf.data` returns just the data of that leaf

Reply:
not sure why a user would require access to the inner implementation. Maybe the API this model exposes is not complete?

Reply:
I've reworked the API surface so that a single `AstarteDataTreeNode<DataType>` interface is exposed to the user, hiding implementation details and differences between leaves and inner nodes.